src/main/java/com/gregtechceu/gtceu/client/renderer/machine/HPCAPartRenderer.java

```diff
@@ -0,0 +1,84 @@
+import net.minecraft.client.renderer.texture.MissingTextureAtlasSprite;
+import net.minecraft.client.resources.model.ModelState;
+import net.minecraft.core.Direction;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.util.RandomSource;
+import org.jetbrains.annotations.Nullable;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNonnullByDefault
+public class HPCAPartRenderer extends TieredHullMachineRenderer {
+
+    private final boolean isAdvanced;
+    private final ResourceLocation texture, activeTexture, activeEmissiveTexture, damagedTexture, damagedActiveTexture, damagedActiveEmissiveTexture;
+
+    public HPCAPartRenderer(boolean isAdvanced, ResourceLocation texture, ResourceLocation damagedTexture) {
+        super(GTValues.ZPM, isAdvanced ? GTCEu.id("block/computer_casing") : GTCEu.id("block/advanced_computer_casing"));
+        this.isAdvanced = isAdvanced;
+        this.texture = texture;
+        this.activeTexture = texture.withSuffix("_active");
+        this.activeEmissiveTexture = activeTexture.withSuffix("_emissive");
+        this.damagedTexture = damagedTexture;
+        this.damagedActiveTexture = damagedTexture.withSuffix("_active");
+        this.damagedActiveEmissiveTexture = damagedActiveTexture.withSuffix("_emissive");
+    }
+
+    @Override
+    public void renderMachine(List<BakedQuad> quads, MachineDefinition definition, @Nullable MetaMachine machine, Direction frontFacing, @Nullable Direction side, RandomSource rand, @Nullable Direction modelFacing, ModelState modelState) {
+        super.renderMachine(quads, definition, machine, frontFacing, side, rand, modelFacing, modelState);
+        if (machine instanceof HPCAComponentPartMachine hpcaComponent) {
+            ResourceLocation texture, emissiveTexture = null;
+            var controller = hpcaComponent.getControllers().isEmpty() ? null : hpcaComponent.getControllers().get(0);
+            if (controller != null && (controller instanceof IWorkable workable && workable.isActive())) {
+                if (hpcaComponent.isDamaged()) {
+                    texture = damagedActiveTexture;
+                    emissiveTexture = damagedActiveEmissiveTexture;
+                } else {
+                    texture = activeTexture;
+                    emissiveTexture = activeEmissiveTexture;
+                }
+            } else {
+                if (hpcaComponent.isDamaged()) {
+                    texture = this.damagedTexture;
+                } else {
+                    texture = this.texture;
+                }
+            }
+            if (ModelFactory.getBlockSprite(texture).atlasLocation().equals(MissingTextureAtlasSprite.getLocation())) {
+                texture = this.texture;
+            }
+            if (texture != null && !ModelFactory.getBlockSprite(texture).atlasLocation().equals(MissingTextureAtlasSprite.getLocation())) {
```

Comment:
Should use else if for avoiding the duplicate lookup:

```suggestion
            } else if (texture != null) {
```

Reply:
no, because the top one reassigns `texture`, and it can still be nonexistent in case of the empty component.